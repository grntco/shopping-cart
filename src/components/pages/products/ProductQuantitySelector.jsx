import styled from 'styled-components'
import Button from '../../reusables/buttons/Button'
import { useState } from 'react'
import PropTypes from 'prop-types'

const ProductQuantitySelector = ({ product, handleAddToCart }) => {
    const [quantity, setQuantity] = useState(1)

    return (
        <StyledQuantitySelector>
            <QuantityInputContainer>
                <Button onClick={handleDecrement}>â€“</Button>
                <input
                    type='number'
                    value={quantity}
                    onChange={handleInputChange}
                    onBlur={(e) => {
                        if (e.target.value === '') setQuantity(1)
                    }}
                />
                <Button onClick={handleIncrement}>+</Button>
            </QuantityInputContainer>
            <Button
                onClick={() => {
                    handleAddToCart(product, quantity)
                }}
            >
                Add to Cart
            </Button>
        </StyledQuantitySelector>
    )

    function handleIncrement() {
        if (quantity < 10) setQuantity(quantity + 1)
    }

    function handleDecrement() {
        if (quantity > 1) setQuantity(quantity - 1)
    }

    function handleInputChange(e) {
        const newValue = e.target.value
        if (newValue <= 10 && newValue >= 0) {
            setQuantity(parseInt(newValue))
        } else if (newValue > 10 || newValue < 0) {
            setQuantity(1)
        } else {
            setQuantity('')
        }
    }
}

const StyledQuantitySelector = styled.div`
    display: flex;
    flex-direction: column;
    gap: 8px;
`

const QuantityInputContainer = styled.div`
    height: 2rem;
    display: flex;
    gap: 2px;

    input {
        width: 2rem;
        text-align: center;
        border: none;
        &::-webkit-inner-spin-button,
        &::-webkit-outer-spin-button {
            -webkit-appearance: none;
            margin: 0;
        }
    }

    Button {
        display: flex;
        align-items: center;
        justify-content: center;
        padding: 8px;
        font-size: 1.2rem;
    }
`

ProductQuantitySelector.propTypes = {
    product: PropTypes.object.isRequired,
    handleAddToCart: PropTypes.func.isRequired,
}

export default ProductQuantitySelector
