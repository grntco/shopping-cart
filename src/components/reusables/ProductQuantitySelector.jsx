import styled from 'styled-components'
import Button from './Button'
import { useState } from 'react'
import PropTypes from 'prop-types'

const ProductQuantitySelector = ({ product, handleAddToCart }) => {
    const [quantity, setQuantity] = useState(1)

    return (
        <StyledQuantitySelector>
            <QuantityInputContainer>
                <Button onClick={handleDecrement}>â€“</Button>
                <input
                    type='number'
                    value={quantity}
                    onChange={handleInputChange}
                />
                <Button onClick={handleIncrement}>+</Button>
            </QuantityInputContainer>
            <Button
                onClick={() => {
                    handleAddToCart(product, quantity)
                }}
            >
                Add to Cart
            </Button>
        </StyledQuantitySelector>
    )

    function handleIncrement() {
        setQuantity(quantity + 1)
    }

    function handleDecrement() {
        if (quantity > 1) setQuantity(quantity - 1)
    }

    function handleInputChange(e) {
        setQuantity(parseInt(e.target.value))
    }
}

const StyledQuantitySelector = styled.div`
    display: flex;
    flex-direction: column;
    gap: 8px;
`

const QuantityInputContainer = styled.div`
    height: 2rem;
    display: flex;
    gap: 2px;

    input {
        width: 2rem;
        text-align: center;
        border: none;
        &::-webkit-inner-spin-button,
        &::-webkit-outer-spin-button {
            -webkit-appearance: none;
            margin: 0;
        }
    }

    Button {
        display: flex;
        align-items: center;
        justify-content: center;
        padding: 8px;
        font-size: 1.2rem;
    }
`

ProductQuantitySelector.propTypes = {
    product: PropTypes.object.isRequired,
    handleAddToCart: PropTypes.func.isRequired,
}

export default ProductQuantitySelector
