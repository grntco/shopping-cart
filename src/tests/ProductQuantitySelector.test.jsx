import { describe, it, expect } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react'
import ProductQuantitySelector from '../components/pages/products/ProductQuantitySelector'
import { BrowserRouter } from 'react-router-dom'
import userEvent from '@testing-library/user-event'

const mockProduct = {
    id: 1,
    title: 'Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops',
    price: 109.95,
    description:
        'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday',
    category: "men's clothing",
    image: 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
    rating: {
        rate: 3.9,
        count: 120,
    },
}

describe('Product Quantity Selector test', () => {
    it('Product Quantity Selector renders correctly', () => {
        render(
            <ProductQuantitySelector
                product={mockProduct}
                handleAddToCart={() => {}}
            />,
            { wrapper: BrowserRouter },
        )
        expect(
            screen.getByTestId('product-quantity-selector'),
        ).toBeInTheDocument()
        expect(
            screen.getByTestId('product-quantity-selector'),
        ).toMatchSnapshot()
    })
})
