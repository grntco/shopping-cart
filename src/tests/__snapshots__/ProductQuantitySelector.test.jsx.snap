// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product Quantity Selector test > Product Quantity Selector renders correctly 1`] = `
<div
  class="sc-guDLey hGbZVM"
  data-testid="product-quantity-selector"
>
  <div
    class="sc-dmyCSP kIuZzL"
  >
    <button
      class="sc-beySPh gepDOA"
    >
      â€“
    </button>
    <input
      type="number"
      value="1"
    />
    <button
      class="sc-beySPh gepDOA"
    >
      +
    </button>
  </div>
  <button
    class="sc-beySPh gepDOA"
  >
    Add to Cart
  </button>
</div>
`;
